presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -226,4 +227,155 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
 
         checkArgument((totalBitLength <= maximumBits), format("The z-address type specified is not large enough to hold %d bits.", totalBitLength));
     }
+
+    /**
+     * Searches for and outputs the addresses within certain ranges in each dimension.
+     *
+     * @param minimumValues the list of minimum values of each dimension
+     * @param maximumValues the list of maximum values of each dimension
+     * @return the list of addresses within the given value ranges
+     */
+    public List<Long> zOrderSearchCurve(List<Integer> minimumValues, List<Integer> maximumValues)
+    {
+        List<Long> addressRanges = zOrderCondensedSearchCurve(minimumValues, maximumValues);
+
+        List<Long> addresses = new ArrayList<>();
+        for (int index = 0; index < addressRanges.size(); index += 2) {
+            long minAddress = addressRanges.get(index);
+            long maxAddress = addressRanges.get(index + 1);
+            for (long address = minAddress; address <= maxAddress; address++) {
+                addresses.add(address);
+            }
+        }
+        return addresses;
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param minimumValues the list of minimum values of each dimension
+     * @param maximumValues the list of maximum values of each dimension
+     * @return the list of address ranges within the given value ranges
+     */
+    // TODO: Create own "pair" object
+    public List<Long> zOrderCondensedSearchCurve(List<Integer> minimumValues, List<Integer> maximumValues)
+    {
+        ZOrderCurveSearcher curveSearcher = new ZOrderCurveSearcher(minimumValues, maximumValues);
+        return curveSearcher.findAddresses();
+    }
+
+    /**
+     * Provides a structure to search for addresses within given value ranges.
+     */
+    private class ZOrderCurveSearcher
```

Comment:
I'm not fully convinced that we need an internal class (e.g. for status tracking etc). It seems like we only call `findAddresses` once here, which could potentially be achieved by helper functions. Am I missing something?

Reply:
I actually ended up creating the `ZOrderCurveSearcher` class as a separate public class that takes in a `ZOrder` variable upon initialization. 
The class has many new variables that are used only for the curve searching function, so I didn't think it was necessary to bunch them into `ZOrder` with the variables used for encoding/decoding.